rpc/dataconn/stream/stream_conn.go

```diff
@@ -143,13 +143,13 @@ func (c *Conn) ReadStream(frameType uint32, closeConnOnClose bool) (_ *StreamRea
 	}
 
 	c.readMtx.Lock()
+	defer c.readMtx.Unlock()
```

Comment:
This unlocks the mutex too early in the case where `c.readClean == true`: the idea is that the goroutine spawned below copies the stream into the io.Pipe and the returned StreamReader reads from it. The Conn is not usable until StreamReader is drained.

So I think we just need to leave the `defer unlock` in the goroutine as it was but add a non-defer unlock in the erorr case where c.readClean == false.

Reply:
Ah, I glanced too quickly and missed the purpose of the defer placement. That was actually my first solution but ended moving the defer, I’ll update the PR 👍